### PR TITLE
Update for recent compiler changes for ref fields

### PR DESCRIPTION
**Pull request by cston**
See https://github.com/dotnet/roslyn/pull/62886.

### CODE REVIEW

**Comment by stephentoub** [comment]
Are all of these changes necessary in order to successfully compile?  If yes, this seems like a fairly significant breaking change, no?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryConverter.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@ internal sealed class IDictionaryConverter<TDictionary>
         : JsonDictionaryConverter<TDictionary, string, object?>
         where TDictionary : IDictionary
     {
-        protected override void Add(string key, in object? value, JsonSerializerOptions options, ref ReadStack state)
+        protected override void Add(string key, in object? value, JsonSerializerOptions options, scoped ref ReadStack state)
```
Comment (jaredpar) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryConverter.cs:19]: Not sure why this one is necessary

**Comment by cston** [comment]
> Are all of these changes necessary in order to successfully compile?

Yes, these are changes needed to compile successfully. These are typically methods that have a `ref` parameter and a `ref struct` parameter.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ArrayConverter.cs [hunk 2]
```diff
@@ -13,13 +13,13 @@ internal sealed class ArrayConverter<TCollection, TElement> : IEnumerableDefault
     {
         internal override bool CanHaveMetadata => false;
 
-        protected override void Add(in TElement value, ref ReadStack state)
+        protected override void Add(in TElement value, scoped ref ReadStack state)
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ArrayConverter.cs:16]: Why is scoped required here?

**Review by stephentoub — COMMENTED** [review]
Other than the comments questioning why certain categories of changes are needed, LGTM.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryConverter.cs (hunk 1 above)
Comment (cston) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryConverter.cs:19]: It was not needed. Removed.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ConcurrentQueueOfTConverter.cs [hunk 3]
```diff
@@ -9,7 +9,7 @@ internal sealed class ConcurrentQueueOfTConverter<TCollection, TElement>
         : IEnumerableDefaultConverter<TCollection, TElement>
         where TCollection : ConcurrentQueue<TElement>
     {
-        protected override void Add(in TElement value, ref ReadStack state)
+        protected override void Add(in TElement value, scoped ref ReadStack state)
```
Comment (jaredpar) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ConcurrentQueueOfTConverter.cs:12]: Why is `scoped` required here?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryOfTKeyTValueConverter.cs [hunk 4]
```diff
@@ -16,7 +16,7 @@ internal sealed class IDictionaryOfTKeyTValueConverter<TDictionary, TKey, TValue
         where TDictionary : IDictionary<TKey, TValue>
         where TKey : notnull
     {
-        protected override void Add(TKey key, in TValue value, JsonSerializerOptions options, ref ReadStack state)
+        protected override void Add(TKey key, in TValue value, JsonSerializerOptions options, scoped ref ReadStack state)
```
Comment (jaredpar) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IDictionaryOfTKeyTValueConverter.cs]: Why is `scoped` needed here?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverter.cs [hunk 5]
```diff
@@ -17,13 +17,13 @@ internal sealed class IEnumerableConverter<TCollection>
     {
         private readonly bool _isDeserializable = typeof(TCollection).IsAssignableFrom(typeof(List<object?>));
 
-        protected override void Add(in object? value, ref ReadStack state)
+        protected override void Add(in object? value, scoped ref ReadStack state)
```
Comment (jaredpar) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverter.cs:20]: Why is `scoped` needed here?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableOfTConverter.cs [hunk 6]
```diff
@@ -15,13 +15,13 @@ internal sealed class IEnumerableOfTConverter<TCollection, TElement>
     {
         private readonly bool _isDeserializable = typeof(TCollection).IsAssignableFrom(typeof(List<TElement>));
 
-        protected override void Add(in TElement value, ref ReadStack state)
+        protected override void Add(in TElement value, scoped ref ReadStack state)
```
Comment (jaredpar) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableOfTConverter.cs:18]: Why is `scoped` needed here?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ConcurrentStackOfTConverter.cs [hunk 7]
```diff
@@ -9,7 +9,7 @@ internal sealed class ConcurrentStackOfTConverter<TCollection, TElement>
         : IEnumerableDefaultConverter<TCollection, TElement>
         where TCollection : ConcurrentStack<TElement>
     {
-        protected override void Add(in TElement value, ref ReadStack state)
+        protected override void Add(in TElement value, scoped ref ReadStack state)
```
Comment (jaredpar) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ConcurrentStackOfTConverter.cs:12]: Why is `scoped` needed here?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ArrayConverter.cs (hunk 2 above)
Comment (cston) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ArrayConverter.cs]: Removed.

**Review by jaredpar — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
I don't see any failures in the arm64 test. Looks like an infrastructure issue.

```xml
[15:05:31.5263400] 2022-08-12 15:05:31.506 System.Runtime.Tests[16226:178107346] Tests run: 49815 Passed: 49582 Inconclusive: 0 Failed: 0 Ignored: 137 Skipped: 96
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**